vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_job.h

```diff
@@ -0,0 +1,74 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_BAT_LEDGER_JOB_H_
+#define BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_BAT_LEDGER_JOB_H_
+
+#include <utility>
+
+#include "base/memory/weak_ptr.h"
+#include "bat/ledger/internal/core/bat_ledger_context.h"
+#include "bat/ledger/internal/core/future.h"
+
+namespace ledger {
+
+// Convenience class for defining job classes that can be started by calling
+// |BATLedgerContext::StartJob|. Subclasses must implement a |Start| method
+// that begins the asynchronous operation.
+//
+// Example:
+//   class MyJob : public BATLedgerJob<int> {
+//    public:
+//     void Start() {
+//       Complete(42);
+//     }
+//   };
+template <typename T>
+class BATLedgerJob : public BATLedgerContext::Object,
+                     public base::SupportsWeakPtr<BATLedgerJob<T>> {
+ public:
+  // Returns the Future for the job.
+  Future<T> GetFuture() { return promise_.GetFuture(); }
+
+ protected:
+  // Completes the job with the specified value.
+  virtual void Complete(T value) { promise_.SetValue(std::move(value)); }
+
+  // Completes the job with the eventual value of the specified future
+  void CompleteWithFuture(Future<T> future) {
+    future.Then(
+        base::BindOnce(&BATLedgerJob::OnInnerCompleted, base::AsWeakPtr(this)));
+  }
+
+  // Returns a |OnceCallback| that wraps the specified member function. The
+  // resulting callback is bound with a WeakPtr for the receiver. It is not
+  // bound with any additional arguments.
+  template <typename Derived, typename... Args>
+  inline static auto ContinueWith(Derived* self, void (Derived::*fn)(Args...)) {
+    return base::BindOnce(fn, base::AsWeakPtr(self));
+  }
+
+  // Returns a lambda that wraps the specified member function. The resulting
+  // lambda will be a no-op if the job has been destroyed. This helper should
+  // only be used for older APIs that require |std::function| callbacks.
+  template <typename Derived, typename... Args>
+  inline static auto CreateLambdaCallback(Derived* self,
```

Comment:
This, too should be called `ContinueWith`, IMO. After all, this function creates a continuation, just as much as `ContinueWith<>()` does.

Reply:
They differ only on return type, so I assume that they'd need different names, but we could do something like `ContinueWithLambda` perhaps. I think I'd prefer to leave this name a bit ugly though, so that usage of `std::function`-based APIs stands out and looks a bit odd - it will nudge us to gradually update those APIs.

Reply:
Sure, your call!

It's quite easy to have the same name for both, though:
```
template <bool legacy = false, typename Derived, typename... Args>
static auto ContinueWith(Derived* self, void (Derived::*fn)(Args...)) {
  if constexpr (legacy) {
    return [weak_self = base::AsWeakPtr(self), fn](auto&&... args) {
      if (weak_self) {
        (weak_self.get()->*fn)(std::forward<decltype(args)>(args)...);
      }
    };
  } else {
    return base::BindOnce(fn, base::AsWeakPtr(self));
  }
}
```
Usage (`legacy` defaults to `false`, whereas the rest of the arguments get deduced, so the above has no impact on default usage):
```
ContinueWith(...);  // base::BindOnce<>()
ContinueWith<true>(...);  // lambda - this usage looks a bit odd, too! :)
```
If you don't like the `bool`, you can also introduce an `enum class` as the non-type template parameter of `ContinueWith<>()`.

Reply:
> If you don't like the bool, you can also introduce an enum class as the non-type template parameter of ContinueWith<>().

I like this option:

```
ContinueWith<ContinuationKind::kLambda>(...);
```

Self-documenting and annoyingly verbose - perfect! 👍

Reply:
This is a great suggestion, but after I made the change above I realized that how these functions are called suggests having the different names.

For the callback-style, we'll typically use it with `Then`:

```
context()
    .Get<T>()
    .DoSomething()
    .Then(ContinueWith(this, ...));
```

which has an intentionally nice English reading, "then continue with..."

For the lambda-style, we'll typically use it in last-arg position:

```
context().GetLedgerImpl()->publisher()->GetServerPublisherInfo(
    "brave.com",
    CreateLambdaCallback(this, ...));
```

where we kind of lose that natural-language flow. Still thinking about this though... 🤔

Reply:
We don't have the natural-language flow in the lambda-style case, regardless of what we call `CreateLambdaCallback<>()`. I'll leave that up to you, though - I'm okay with both! 🙂

Reply:
I decided to go with `ContinueWithLambda`, which seems a good mix of

* Obvious and easily searchable.
* Not too verbose.
* A simple extension of the `ContinueWith` name, so users can extrapolate the meaning.

This has been really helpful - thanks!